### PR TITLE
Worker support for multiple webUIs

### PR DESCRIPTION
**Pull request by aaannz**
Allows worker to register and work on jobs from multiple openQA servers.

Hosts to register with are configured in `workers.ini` in global section `HOST` key. Accepts space separated hostnames. Optionally specific share directory can be configured in respective host sections. See example in `workers.ini`. If share directory is not configured, worker will check if `/var/lib/openqa/$hostname` does not exist or use default `/var/lib/openqa/share`.

There are also changes in `api_call` method, now accepts named options `json, params, callback`. Callback is used to queue actions dependent on result of api call.

### CODE REVIEW

File: lib/OpenQA/Utils.pm [hunk 1]
```diff
@@ -101,12 +100,29 @@ sub testcasedir($$) {
...
+        if (-d $d) {
+            $dir = $d;
+        }
+    }
     $dir .= "-$version" if $version && -e "$dir-$version";
 
     return $dir;
 }
 
+# Call this when $prjdir is changed to reevalue all dependent directories
```
Comment (okurz) [inline lib/OpenQA/Utils.pm]: s/reevalue/re-evaluate/?

File: lib/OpenQA/Worker.pm [hunk 2]
```diff
@@ -29,24 +31,44 @@ use OpenQA::Worker::Pool qw(lockit clean_pool);
...
+        my $dir;
+        for my $d (@dirs) {
+            next unless $d;
+            if (-d $d) {
+                $dir = $d;
+                last;
+            }
+        }
+        unless ($dir) {
+            print STDERR "Can not find working directory for host $h. Ignoring host\n";
```
Comment (okurz) [inline lib/OpenQA/Worker.pm]: better should use the logging functions, e.g. log_err

File: lib/OpenQA/Worker/Jobs.pm [hunk 3]
```diff
@@ -91,27 +91,37 @@ sub _kill_worker($) {
...
+            $job = $res->{job};
+            if ($job && $job->{id}) {
+                Mojo::IOLoop->next_tick(sub { start_job($host) });
+            }
+            else {
+                $job = undef;
+                Mojo::IOLoop->next_tick(sub { check_job(@todo) });
+            }
+            $check_job_running->{$host} = 0;
+        });
```
Comment (okurz) [inline lib/OpenQA/Worker/Jobs.pm]: don't agree here but with all other previous comments :-)

Comment (kraih) [inline lib/OpenQA/Worker/Jobs.pm]: At the very least an empty line after `my (@todo) = @_;`, and one  before the `api_call`. 😉

**Comment by aaannz** [comment]
Good. Existing tests pass, now to write new ones.

**Comment by aaannz** [comment]
What does the codecov/patch means?

**Comment by kraih** [comment]
Found this. https://github.com/codecov/support/wiki/Patch-Status

**Comment by coolo** [comment]
Now please rebase - the broken review on patch is based on the 'there can never be too much coverage' approach :)

**Comment by aaannz** [comment]
Rebased, seems to me that the uncovered lines are handlers for exceptions. That would be nice to test, but I need to mock so much for that I'll need to figure out some better way.

**Comment by coolo** [comment]
yeah, once the updated codecov report is in I'll have a 2nd look and then merge if I agree with your assessment. The changes and patch review are merely to catch regressions in smaller changes. For larger changes they are rather fragile

**Comment by coolo** [comment]
I'm building a test package to deploy on my staging server.

**Comment by coolo** [comment]
OK, no luck with that 

```
[   69s] ./script/generate-packed-assets
[   70s] Uncaught exception from user code:
[   70s]        org.freedesktop.DBus.Error.FileNotFound: Failed to connect to socket /run/dbus/system_bus_socket: No such file or directory
[   70s]        Net::DBus::Binding::Bus::new('Net::DBus::Binding::Bus', 'type', 1) called at /usr/lib/perl5/vendor_perl/5.18.2/x86_64-linux-thread-multi/Net/DBus.pm line 178
[   70s]        Net::DBus::system('Net::DBus') called at /usr/lib/perl5/vendor_perl/5.18.2/x86_64-linux-thread-multi/Net/DBus.pm line 153
[   70s]        Net::DBus::find('Net::DBus', 'nomainloop', '') called at /home/abuild/rpmbuild/BUILD/openQA-4.4.1484756717.5937a29f/script/../lib/OpenQA/IPC.pm line 57
[   70s]        OpenQA::IPC::new('OpenQA::IPC', '', 1) called at /home/abuild/rpmbuild/BUILD/openQA-4.4.1484756717.5937a29f/script/../lib/OpenQA/IPC.pm line 65
[   70s]        OpenQA::IPC::ipc('OpenQA::IPC', '', 1) called at /home/abuild/rpmbuild/BUILD/openQA-4.4.1484756717.5937a29f/script/../lib/OpenQA/WebAPI.pm line 483
[   70s]        OpenQA::WebAPI::startup('OpenQA::WebAPI=HASH(0x1599f10)') called at /usr/lib/perl5/vendor_perl/5.18.2/Mojolicious.pm line 174
[   70s]        Mojolicious::new('OpenQA::WebAPI') called at /usr/lib/perl5/vendor_perl/5.18.2/Mojo/Server.pm line 17
[   70s]        Mojo::Server::build_app('Mojo::Server=HASH(0x146ee18)', 'OpenQA::WebAPI') called at /usr/lib/perl5/vendor_perl/5.18.2/Mojolicious/Commands.pm line 71
[   70s]        Mojolicious::Commands::start_app('Mojolicious::Commands', 'OpenQA::WebAPI') called at /home/abuild/rpmbuild/BUILD/openQA-4.4.1484756717.5937a29f/script/../lib/OpenQA/WebAPI.pm line 502
[   70s]        OpenQA::WebAPI::run() called at ./script/openqa line 35
[   70s] Makefile:8: recipe for target 'install' failed
```

Should I set OPENQA_IPC_TEST for that? (I wonder it worked before)

**Comment by AdamWill** [comment]
This actually causes a rather large problem: occasionally, it kills workers. I noticed recently several of our workers have died with logs like this:

    Mar 09 20:08:36 qa09.qa.fedoraproject.org worker[5892]: [INFO] 42187: WORKING 77146
    Mar 09 20:22:56 qa09.qa.fedoraproject.org worker[5892]: [ERROR] 404 response: Not Found (remaining tries: 0)
    Mar 09 20:22:56 qa09.qa.fedoraproject.org worker[5892]: [ERROR] Job aborted because web UI doesn't accept updates anymore (likely considers this job dead)
    Mar 09 20:22:57 qa09.qa.fedoraproject.org worker[5892]: Mojo::Reactor::Poll: Timer failed: No worker id or webui host set! at /usr/share/openqa/script/../lib/OpenQA/Worker/Common.pm line 181.
    Mar 09 20:23:06 qa09.qa.fedoraproject.org worker[5892]: WebUI Mojo::IOLoop=HASH(0x1768308) is unknown! - Should not happen but happened, exiting! at /usr/share/openqa/script/../lib/OpenQA/Worker/Common.pm line 404.
    Mar 09 20:23:06 qa09.qa.fedoraproject.org worker[5892]: [INFO] registering worker with openQA Mojo::IOLoop=HASH(0x1768308)...
    Mar 09 20:23:28 qa09.qa.fedoraproject.org systemd[1]: openqa-worker@4.service: Killing process 42187 (perl) with signal SIGKILL.
    Mar 09 20:23:28 qa09.qa.fedoraproject.org systemd[1]: openqa-worker@4.service: Killing process 42224 (/usr/bin/isotov) with signal SIGKILL.
    Mar 09 20:23:28 qa09.qa.fedoraproject.org systemd[1]: openqa-worker@4.service: Killing process 42227 (/usr/bin/isotov) with signal SIGKILL.
    Mar 09 20:23:28 qa09.qa.fedoraproject.org systemd[1]: openqa-worker@4.service: Killing process 42232 (/usr/bin/isotov) with signal SIGKILL.
    Mar 09 20:23:28 qa09.qa.fedoraproject.org systemd[1]: openqa-worker@4.service: Killing process 42239 (videoencoder) with signal SIGKILL.
    Mar 09 20:23:28 qa09.qa.fedoraproject.org systemd[1]: openqa-worker@4.service: Killing process 42257 (qemu-system-x86) with signal SIGKILL.

I believe this problem is caused by the way this commit changes the behaviour of `register_worker`, interacting with a call to that function in `api_call`.

`api_call` has these lines, which kick in when an API call attempt failed:

        if (!$tries) {
            # abort the current job, we're in trouble - but keep running to grab the next
            OpenQA::Worker::Jobs::stop_job('api-failure');
            $hosts->{$host}{workerid} = undef;
            $host = undef;
            add_timer('register_worker', 10, \&register_worker, 1);
            $callback->();
            return;
        }

Note the way it calls `register_worker`: it doesn't specify any variables. Particularly, it doesn't specify a host name, and in fact, given the way the code flows, it can't, because it just got done *un*setting `$host`. It seems to have an assumption that unsetting `$host` and then calling `register_worker` (OK, setting a timer to do it a bit later) should work.

After this PR, though, that *doesn't* work, because `register_worker` bails if it isn't passed a correct hostname.

It looks like what's happening is that somehow the IO loop object itself winds up getting passed as the first arg to `register_worker` by this code, and `register_worker` obviously doesn't find it as a known host, and bails.

The 'naive' fix would be to move the `add_timer` call up one line and make it the same as the other two `register_worker` timer lines in the code:

            add_timer('register_worker', 10, sub { register_worker($host) }, 1);

but I'm not sure if that is correct in this context.

**Comment by aaannz** [comment]
Yes, hence this request https://github.com/os-autoinst/openQA/pull/1215 you can try. I'm stuck on test for it for way longer than acceptable, unfortunately. And https://progress.opensuse.org/issues/16380

**Comment by AdamWill** [comment]
OK, so: *why* did this PR (specifically, commit 9261f7af1b2751eeca1fa0194a16ec6dad9c85b1 , "Adapt jobs helpers to non recursive api_call and multihost") remove the `$update_status_running = 0;` line from the end of `update_status`, in `Worker/Jobs.pm`? I cannot see a justification for this and I'm kinda suspicious of it. I'm *especially* suspicious of it because I can't find *anywhere* that `$update_status_running` gets set back to 0, if we have a status to update and we're doing it over websockets: the `upload_status` sub sets it to 0 in several cases ('no tests scheduled', 'no status to update', 'status updated via API call') but it does *not* - so far as I can tell - set it to 0 when there is a status to update and it is updated via websockets rather than via API call (as determined by the `WORKER_USE_WEBSOCKETS` env var).

**Comment by aaannz** [comment]
That specific commit also adds `$update_status_running = 0` to `upload_status`. It removes it because at that time, `update_status` returned without blocking and `$update_status_running` was reset in callbacks. If it is not there in current code, that's the problem of different PRs.

**Comment by AdamWill** [comment]
It added it on *some* paths through `upload_status`, but not *all* paths. It didn't add it on the websockets path (which already existed at the time), as I noted. It also didn't add it to some very early failure paths, though I don't know how much of a problem that is.

**Pull request merged**